<a href="https://colab.research.google.com/github/anumit-web/python-interview-2024/blob/main/machine%20learning/Unsupervised_Learning_clustering_type_iris_flower_clustering_Gaussian_mixture_model_(GMM).ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

# Python interview preparation

# Machine Learning

# Reinforcement Learning (RL)

# Chapter 1

---

---

# (Reinforcement Learning (RL) definition from Google

## Reinforcement Learning 101.

Reinforcement learning (RL) is **a machine learning technique that teaches software to make decisions to achieve optimal results.** It's a powerful method for AI systems to learn to achieve optimal outcomes in environments they haven't seen before



**Here are some key characteristics of RL:**

**Trial-and-error:** RL mimics the way humans learn by trial and error to achieve their goals.

**Reward-and-punishment:** RL uses a reward-and-punishment paradigm to process data.

**Delayed gratification:** RL algorithms can consider delayed rewards, as the best strategy may require short-term sacrifices.

**No human guidance:** RL agents learn to perform tasks without any guidance from a human user.

**Goal-oriented:** RL is task-oriented, with the goal of maximizing the reward over a given time period.








1. Text analysis
2. Text mining

Text analysis = the process of analyzing and extracting meaningful insights from unstructured text data

---


## **What is Sentiment Analysis?**

sentiment analysis = determining the emotional tone of the text.

subfields of text analysis is sentiment analysis


## **What is the use of Sentiment Analysis?**

1. customer feedback analysis
2. social media monitoring
2. market research
2. determine the emotional tone or sentiment expressed in a text
3. positive, negative, or neutral

## **What are the Three Methodologies for Sentiment Analysis?**

1. Lexicon-based analysis approach
2. Machine learning (ML) approach
3. Pre-trained transformer-based deep learning



1. **Lexicon-based analysis approach** - is a straightforward approach to sentiment analysis, but it may not be as accurate as more complex methods.
2. **Machine learning (ML) approach** - approaches can be more accurate, but they require labeled training data and may be more computationally expensive.
3. **Pre-trained transformer-based deep learning** - approaches can achieve state-of-the-art accuracy but require significant computational resources and may not be practical for all use cases.






---

In [ ]:
import numpy as np
import pandas as pd

In [ ]:
import sys

print(sys.version)

3.10.12 (main, Sep 11 2024, 15:47:36) [GCC 11.4.0]


In [ ]:
print("Hello, World!")

Hello, World!


# Python code

In [ ]:
import nltk

nltk.download('all')

[nltk_data] Downloading collection 'all'
[nltk_data]    | 
[nltk_data]    | Downloading package abc to /root/nltk_data...
[nltk_data]    |   Package abc is already up-to-date!
[nltk_data]    | Downloading package alpino to /root/nltk_data...
[nltk_data]    |   Package alpino is already up-to-date!
[nltk_data]    | Downloading package averaged_perceptron_tagger to
[nltk_data]    |     /root/nltk_data...
[nltk_data]    |   Package averaged_perceptron_tagger is already up-
[nltk_data]    |       to-date!
[nltk_data]    | Downloading package averaged_perceptron_tagger_eng to
[nltk_data]    |     /root/nltk_data...
[nltk_data]    |   Package averaged_perceptron_tagger_eng is already
[nltk_data]    |       up-to-date!
[nltk_data]    | Downloading package averaged_perceptron_tagger_ru to
[nltk_data]    |     /root/nltk_data...
[nltk_data]    |   Package averaged_perceptron_tagger_ru is already
[nltk_data]    |       up-to-date!
[nltk_data]    | Downloading package averaged_perceptron_tagger_r

True

Step 1 - Import libraries and load dataset

In [ ]:
# import libraries
import pandas as pd

import nltk

from nltk.sentiment.vader import SentimentIntensityAnalyzer

from nltk.corpus import stopwords

from nltk.tokenize import word_tokenize

from nltk.stem import WordNetLemmatizer


# download nltk corpus (first time only)
import nltk

nltk.download('all')




# Load the amazon review dataset

df = pd.read_csv('https://raw.githubusercontent.com/pycaret/pycaret/master/datasets/amazon.csv')

df

[nltk_data] Downloading collection 'all'
[nltk_data]    | 
[nltk_data]    | Downloading package abc to /root/nltk_data...
[nltk_data]    |   Package abc is already up-to-date!
[nltk_data]    | Downloading package alpino to /root/nltk_data...
[nltk_data]    |   Package alpino is already up-to-date!
[nltk_data]    | Downloading package averaged_perceptron_tagger to
[nltk_data]    |     /root/nltk_data...
[nltk_data]    |   Package averaged_perceptron_tagger is already up-
[nltk_data]    |       to-date!
[nltk_data]    | Downloading package averaged_perceptron_tagger_eng to
[nltk_data]    |     /root/nltk_data...
[nltk_data]    |   Package averaged_perceptron_tagger_eng is already
[nltk_data]    |       up-to-date!
[nltk_data]    | Downloading package averaged_perceptron_tagger_ru to
[nltk_data]    |     /root/nltk_data...
[nltk_data]    |   Package averaged_perceptron_tagger_ru is already
[nltk_data]    |       up-to-date!
[nltk_data]    | Downloading package averaged_perceptron_tagger_r

reviewText  Positive
0      This is a one of the best apps acording to a b...         1
1      This is a pretty good version of the game for ...         1
2      this is a really cool game. there are a bunch ...         1
3      This is a silly game and can be frustrating, b...         1
4      This is a terrific game on any pad. Hrs of fun...         1
...                                                  ...       ...
19995  this app is fricken stupid.it froze on the kin...         0
19996  Please add me!!!!! I need neighbors! Ginger101...         1
19997  love it!  this game. is awesome. wish it had m...         1
19998  I love love love this app on my side of fashio...         1
19999  This game is a rip off. Here is a list of thin...         0

[20000 rows x 2 columns]

Step 2 - Preprocess text

In [ ]:
# create preprocess_text function
def preprocess_text(text):

    # Tokenize the text

    tokens = word_tokenize(text.lower())




    # Remove stop words

    filtered_tokens = [token for token in tokens if token not in stopwords.words('english')]




    # Lemmatize the tokens

    lemmatizer = WordNetLemmatizer()

    lemmatized_tokens = [lemmatizer.lemmatize(token) for token in filtered_tokens]




    # Join the tokens back into a string

    processed_text = ' '.join(lemmatized_tokens)

    return processed_text

# apply the function df

df['reviewText'] = df['reviewText'].apply(preprocess_text)
df

reviewText  Positive
0      one best apps acording bunch people agree bomb...         1
1      pretty good version game free . lot different ...         1
2      really cool game . bunch level find golden egg...         1
3      silly game frustrating , lot fun definitely re...         1
4      terrific game pad . hr fun . grandkids love . ...         1
...                                                  ...       ...
19995  app fricken stupid.it froze kindle wont allow ...         0
19996  please add ! ! ! ! ! need neighbor ! ginger101...         1
19997  love ! game . awesome . wish free stuff house ...         1
19998  love love love app side fashion story fight wo...         1
19999  game rip . list thing make better & bull ; fir...         0

[20000 rows x 2 columns]

Step 3 - NLTK Sentiment Analyzer

In [ ]:
# initialize NLTK sentiment analyzer

analyzer = SentimentIntensityAnalyzer()


# create get_sentiment function

def get_sentiment(text):

    scores = analyzer.polarity_scores(text)

    sentiment = 1 if scores['pos'] > 0 else 0

    return sentiment




# apply get_sentiment function

df['sentiment'] = df['reviewText'].apply(get_sentiment)

df

reviewText  Positive  sentiment
0      one best apps acording bunch people agree bomb...         1          1
1      pretty good version game free . lot different ...         1          1
2      really cool game . bunch level find golden egg...         1          1
3      silly game frustrating , lot fun definitely re...         1          1
4      terrific game pad . hr fun . grandkids love . ...         1          1
...                                                  ...       ...        ...
19995  app fricken stupid.it froze kindle wont allow ...         0          0
19996  please add ! ! ! ! ! need neighbor ! ginger101...         1          1
19997  love ! game . awesome . wish free stuff house ...         1          1
19998  love love love app side fashion story fight wo...         1          1
19999  game rip . list thing make better & bull ; fir...         0          1

[20000 rows x 3 columns]

In [ ]:
from sklearn.metrics import confusion_matrix

print(confusion_matrix(df['Positive'], df['sentiment']))

[[ 1131  3636]
 [  576 14657]]


In [ ]:
from sklearn.metrics import classification_report

print(classification_report(df['Positive'], df['sentiment']))

              precision    recall  f1-score   support

           0       0.66      0.24      0.35      4767
           1       0.80      0.96      0.87     15233

    accuracy                           0.79     20000
   macro avg       0.73      0.60      0.61     20000
weighted avg       0.77      0.79      0.75     20000



As you can see, the overall accuracy of this rule-based sentiment analysis model is 79%. Since this is labeled data, you can also try to build a ML model to evaluate if an ML-based approach will result in better accuracy.

[University of Pennsylvania] (https://guides.library.upenn.edu/penntdm/python/nltk)

[NLTK Sentiment Analysis Tutorial for Beginners ] ( https://www.datacamp.com/tutorial/text-analytics-beginners-nltk )

[Sentiment Analysis: First Steps With Python's NLTK Library] (https://realpython.com/python-nltk-sentiment-analysis/)
